proto/backend.proto

```diff
@@ -306,3 +306,160 @@ message StreamPacket {
+  AppInstanceSettings appInstanceSettings = 4;
+
+  // Valid datasource instance settings if they exist in the request
+  DataSourceInstanceSettings dataSourceInstanceSettings = 5;
+}
+
+// AdmissionRequest contains information from a kubernetes Admission request and decoded object(s).
+// See: https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/admission/types.go#L41
+// And: https://github.com/grafana/grafana-app-sdk/blob/main/resource/admission.go#L14
+// NOTE: this does not include a plugin context
+message AdmissionRequest {
+  // NOTE: this may not include app or datasource instance settings depending on the request
+  PluginContext pluginContext = 1;
+
+  // The requested operation
+  StorageOperation operation = 2;
+  
+  // Object is the object in the request.  This includes the full metadata envelope.
+  // The Group+Version+Kind will be included in the payload
+  bytes object_bytes = 3;
+
+  // OldObject is the object as it currently exists in storage. This includes the full metadata envelope.
+  bytes old_object_bytes = 4;
+}
+
+// ConversionRequest supports converting an object from on version to another 
+message ConversionRequest {
+  // NOTE: this context exists because most of the middleware depends on it
+  // The format conversions most likely do not depend on instance settings passed with the context
+  PluginContext pluginContext = 1;
+
+  enum ObjectEnvelope {
+    // Kubernetes style resource
+    RESOURCE = 0;
+    // Full query request
+    QUERY_DATA_REQUEST = 1;
+  }
+
+  // object metadata wrapper
+  ObjectEnvelope envelope = 2;
+  
+  // Object is the object in the request.  This includes the full metadata envelope.
+  bytes object_bytes = 3;
+
+  // Target converted version
+  string target_version = 4;
+}
+
+// Operation is the type of resource operation being checked for admission control
+// https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/admission/types.go#L158
+enum StorageOperation {
+  CREATE = 0;
+  UPDATE = 1;
+  DELETE = 2;
+}
+
+// See https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/admission/types.go#L118
+message StorageResponse {
+  // Allowed indicates whether or not the admission request was permitted.
+  bool allowed = 1;
+
+  // Result contains extra details into why an admission request was denied.
+  // This field IS NOT consulted in any way if "Allowed" is "true".
+  // +optional
+  StatusResult result = 2;
+
+  // AuditAnnotations is an unstructured key value map set by remote admission controller (e.g. error=image-blacklisted).
+  // MutatingAdmissionWebhook and ValidatingAdmissionWebhook admission controller will prefix the keys with
+  // admission webhook name (e.g. imagepolicy.example.com/error=image-blacklisted). AuditAnnotations will be provided by
+  // the admission webhook to add additional context to the audit log for this request.
+  // +optional
+  map<string,string> auditAnnotations = 3;
```

Comment:
Are we guaranteed that annotations can't have repeating keys?

Reply:
Copied directly from: https://github.com/kubernetes/kubernetes/blob/v1.30.0/pkg/apis/admission/types.go#L140